source/mir/algorithm/iteration.d

```diff
@@ -3891,3 +3892,133 @@ version(mir_test)
     auto rc = matrix.byDim!1.map!filterPositive;
     assert(equal!equal(rc, [ [3, 100], [], [400, 102] ]));
 }
+
+/++
+Implements the homonym function (also known as `accumulate`, $(D
+compress), `inject`, or `foldl`) present in various programming
+languages of functional flavor. The call `fold!(fun)(slice, seed)`
+first assigns `seed` to an internal variable `result`,
+also called the accumulator. Then, for each element `x` in $(D
+slice), `result = fun(result, x)` gets evaluated. Finally, $(D
+result) is returned. The one-argument version `fold!(fun)(slice)`
+works similarly, but it uses the first element of the slice as the
+seed (the slice must be non-empty).
+
+Params:
+    fun = the predicate function to apply to the elements
+
+See_Also:
+    $(HTTP en.wikipedia.org/wiki/Fold_(higher-order_function), Fold (higher-order function))
+
+    $(LREF sum) is similar to `fold!((a, b) => a + b)` that offers
+    precise summing of floating point numbers.
+
+    This is functionally equivalent to $(LREF reduce) with the argument order
+    reversed, and without the need to use $(REF_ALTTEXT `tuple`,tuple,std,typecons)
+    for multiple seeds.
++/
+template fold(alias fun)
+{
+    /++
+    Params:
+        slice = A slice, range, and array.
+        seed = An initial accumulation value (optional).
+    Returns:
+        the accumulated result
+    +/
+    @optmath auto fold(Slice, S)(scope Slice slice, S seed)
+    {
+        return reduce!fun(seed, slice);
+    }
+    
+    @optmath auto fold(Slice)(scope Slice slice)
+        if (isIterable!Slice)
```

Comment:
Can we remove the API without seed?
It will cause people to write slow code, `flattened` returns a slower range for Canonical and Universal tensors.
In addition, it should throw an exception in case of an empty input.

Reply:
@9il The way I set it up for the assert produces a good error message when calling it for an empty array with `arr.fold!((a, b) => a)(0)`, but not for `arr.fold!(a => a)(0)` because the call to `reduce` does not compile properly and generates an error before the assert gets run.